### ISSUE TITLE
Compatibility with 2025 Windows redesign

### ISSUE THREAD

**Issue by Aldaviva** [Aldaviva/AuthenticatorChooser#9]
https://blogs.windows.com/windowsdeveloper/2024/10/08/passkeys-on-windows-authenticate-seamlessly-with-passkey-providers/

First available in Insider Beta channel
https://blogs.windows.com/windows-insider/2024/11/01/announcing-windows-11-insider-preview-build-22635-4440-beta-channel/

Will be in Windows 11 25H2
https://blogs.windows.com/windows-insider/2025/06/27/announcing-windows-11-insider-preview-build-26200-5670-dev-channel/#:~:text=Passwordless%20Improvements

Released in Windows 11 24H2 August 2025 Preview Update with gradual rollout
https://support.microsoft.com/en-us/topic/august-29-2025-kb5064081-os-build-26100-5074-preview-3f9eb9e1-72ca-4b42-af97-39aace788d93

![](https://blogs.windows.com/wp-content/uploads/prod/sites/44/2024/11/Passkey-Sign-in-CredUIRejuv.png)

# Differences that require updates
- [x] "Sign in with your passkey" text changed to "Choose a passkey", although other text appears unchanged so far. Seems like this comes from `webauthn.dll.mui` in string table 67, string 1057 (although this file no longer shows as loaded by `CredentialUIBroker.exe`).
- [x] `CREDENTIALS_LIST_ID_CONDITION` element no longer has an automation ID, but its class name is "ListView" (the only such child of the `outerScrollViewer`).
- [x] There is no Next button, and instead the wizard advances to the next page when you select the list item. This means the check to prevent advancing on <kbd>Shift</kbd> or if there are other choices needs to be made earlier in 25H2, before selecting the Security key choice.
- [x] The way you use a security key even when a valid passkey is found in the TPM is no longer a list item, it is now a separate button with different text, "Choose a different passkey". This should only be clicked when the third child of the `outerScrollViewer` does not have text "Security key" (which technically could be confused for a username of a TPM passkey).

**Comment by felipecrs** [comment]
I can confirm it's not working with Windows 11 25H2 (Insider Beta Channel).

<img width="691" height="625" alt="Image" src="https://github.com/user-attachments/assets/1d7b5b8b-7519-4d88-b095-c150c0026675" />

**Comment by Aldaviva** [comment]
Last time I tested on Insider Beta it had been reverted to a branch with the old 24H2 behavior, so it still worked for me. Thanks for letting me know about that channel being changed, I'll try updating and seeing how it works currently.

**Comment by Aldaviva** [comment]
Checking if this is included in the [2025-08 Cumulative Update Preview for Windows 11 24H2 (KB5064081, 26100.5074)](https://support.microsoft.com/en-us/topic/august-29-2025-kb5064081-os-build-26100-5074-preview-3f9eb9e1-72ca-4b42-af97-39aace788d93)…

**Comment by Aldaviva** [comment]
I have been using [Inspect](https://learn.microsoft.com/en-us/windows/win32/winauto/inspect-objects?redirectedfrom=MSDN) and found it pretty helpful because it shows the UI Automation tree and has a follow-the-mouse feature.

**Comment by Aldaviva** [comment]
Sure, feel free to send some screenshots of that to help as a starting point for this change.

**Comment by godeater** [comment]
Top level windows security window:
<img width="2254" height="1496" alt="Image" src="https://github.com/user-attachments/assets/a373b1b7-e9ac-4a65-9683-9df0a379a67e" />

List object:

<img width="2254" height="1496" alt="Image" src="https://github.com/user-attachments/assets/be370c29-5c15-4a1e-b32e-4be7130634bb" />

List item object:

<img width="2254" height="1496" alt="Image" src="https://github.com/user-attachments/assets/2a486e99-516b-4196-a6fd-9469515e5a3e" />

**Comment by Aldaviva** [comment]
Thanks, looks good so far.

**Comment by godeater** [comment]
Was just looking through the code a bit more, and I'm guessing now that the issue might be down to all the strings you're checking for in the window (The "Making sure it's you" stuff etc.), which don't seem to be present on the new UI.

**Comment by godeater** [comment]
Yeah, I think the problem is those strings aren't there at ALL in the new UI.

The first window you see is this one:

<img width="570" height="356" alt="Image" src="https://github.com/user-attachments/assets/3d4febdd-9196-473b-919b-419391f75691" />

And then this one after you select "Security Key"

<img width="570" height="572" alt="Image" src="https://github.com/user-attachments/assets/97d55754-35c5-4e54-8ebd-da81f2aef60b" />

**Comment by Aldaviva** [comment]
According to [Available languages for Windows](https://learn.microsoft.com/en-us/windows-hardware/manufacture/desktop/available-language-packs-for-windows?view=windows-11), those are the only two English language packs. It's possible that list is wrong, but I don't yet know of a better source of language codes to pass to `Install-Language`.

**Comment by godeater** [comment]
Nope - seems like you're correct. I got a colleague to run `Get-InstalledLanguage` on his laptop, and this is what he got back:

```
Language Language Packs  Language Features
-------- --------------  -----------------
en-AU    None            BasicTyping, Speech, TextToSpeech
en-GB    LpCab, LXP      BasicTyping, Handwriting, Speech, TextToSpeech, OCR
en-US    LpCab           BasicTyping, Handwriting, Speech, TextToSpeech, OCR
```

No language pack for en-AU, and I presume the other variants are the same. Anyway. Enough off topic!

**Comment by Aldaviva** [comment]
This seems to be working in both 24H2 and 25H2 after some brief testing. Anyone can feel free to try out [**this build**](https://github.com/user-attachments/files/22398895/AuthenticatorChooser-0.4.0-alpha.1.zip) (or compile from the `25h2` branch). If you encounter any issues, let me know, and remember to [enable the log file](https://github.com/Aldaviva/AuthenticatorChooser/wiki/Command%E2%80%90line-arguments#--logfilename) and upload it too.

**Comment by felipecrs** [comment]
I can confirm it's working for me!

**Comment by godeater** [comment]
> This seems to be working in both 24H2 and 25H2 after some brief testing. Anyone can feel free to try out [**this build**](https://github.com/user-attachments/files/22398895/AuthenticatorChooser-0.4.0-alpha.1.zip) (or compile from the `25h2` branch). If you encounter any issues, let me know, and remember to [enable the log file](https://github.com/Aldaviva/AuthenticatorChooser/wiki/Command%E2%80%90line-arguments#--logfilename) and upload it too.

Working fine here too - thank you very much! (24H2 here)

**Comment by Aldaviva** [comment]
Thanks very much for testing, @felipecrs and @godeater. I have created a [beta release](https://github.com/Aldaviva/AuthenticatorChooser/releases/tag/0.4.0-beta.1), and will release it as stable version if no issues appear soon.

**Comment by godeater** [comment]
Thank _you_ for such a quick re-implemetation!

**Comment by rockerest** [comment]
Beta is working beautifully for me so far!

**Massive thank you** for continuing to support this.
I'd be happy to throw a couple bucks your way on Ko-Fi or something if you want it; this is a valuable service to me!

**Comment by Aldaviva** [comment]
I'm really glad it's helpful for people! Donations are never required but always appreciated; I currently only have [PayPal](https://paypal.me/aldaviva).

**Comment by Aldaviva** [comment]
No issues found so far, therefore I am promoting the 0.4.0-beta.1 release to [0.4.0](https://github.com/Aldaviva/AuthenticatorChooser/releases/tag/0.4.0) stable. It's the same artifact, so anyone already using that beta has no reason to re-download the EXE.
